xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-index/src/main/java/org/xwiki/extension/index/internal/ExtensionInstallListener.java

```diff
@@ -72,15 +74,32 @@ public void onEvent(Event event, Object source, Object data)
         // TODO: make that asynchronous
         try {
             if (event instanceof ExtensionUninstalledEvent) {
-                onUninstalled(((ExtensionEvent) event).getExtensionId(), (((ExtensionEvent) event).getNamespace()));
-            } else if (event instanceof ExtensionInstalledEvent || event instanceof ExtensionUpgradedEvent) {
-                onInstalled(((ExtensionEvent) event).getExtensionId(), (((ExtensionEvent) event).getNamespace()));
+                onUninstalled(((ExtensionEvent) event).getExtensionId(), ((ExtensionEvent) event).getNamespace());
+            } else if (event instanceof ExtensionInstalledEvent) {
+                onInstalled(((ExtensionEvent) event).getExtensionId(), ((ExtensionEvent) event).getNamespace());
+            } else if (event instanceof ExtensionUpgradedEvent) {
+                ExtensionUpgradedEvent extensionUpgradedEvent = (ExtensionUpgradedEvent) event;
+                String namespace = extensionUpgradedEvent.getNamespace();
+                onUpgraded(extensionUpgradedEvent.getExtensionId(), ((Collection<InstalledExtension>) data), namespace);
             }
         } catch (Exception e) {
             this.logger.error("Failed to update the local extension store", e);
         }
     }
 
+    private void onUpgraded(ExtensionId extensionId, Collection<InstalledExtension> previousExtensions,
+        String namespace) throws SolrServerException, IOException
+    {
+        for (InstalledExtension previousExtension : previousExtensions) {
+            if (previousExtension.getNamespaces().contains(namespace)) {
```

Comment:
I don't understand why you check this:
* the extension cannot exist in the same namespace if it's listed as previous extension
* this logic does not work with the root (`null`) namespace anyway, since the list is empty in that case

Reply:
Removed, in indeed realized in the meantime that I was causing NPEs